File: binary-search/69.py
class Solution:
    def mySqrt(self, x: int) -> int:
        if x in [0, 1]: return x
        return self.help_method(x, 0, x)

    def help_method(self, x, left, right):
        if left ** 2 == x:
            return left
        if right == left + 1 and left ** 2 < x and right ** 2 > x:
            return left
        mid = int((left + right) / 2)
        return self.help_method(x, mid, right) if mid ** 2 <= x else self.help_method(x, left, mid)
def main():
    print(Solution().mySqrt(9))

if __name__ == "__main__":
    main()

File: leetcode-cpp/69.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int mySqrt(int x) {
        if (x == 0 || x == 1) return x;
        long long left = 1, right = x;
        while(1){
            if (right == left + 1 && left * left < x && right * right > x)
                return (int)left;
            long long mid = (left + right) / 2;
            long long temp = mid * mid;
            if (temp == x) return mid;
            else if (temp < x) left = mid;
            else right = mid;
        }
    }
};

int main(){
    Solution a;
    for (int x = 0; x <= 30; x++)
        cout << x << " " << a.mySqrt(x) << endl;
    return 0;
}
